Improvements to pattern_matching::myers (generic bit vectors, ambiguities and traceback)

markschl opened the pull request:
This pull request makes the `myers` module more flexible and adds functions for obtaining a traceback at any position. The existing methods are not changed and performance should not be different. Thinking of a good API was harder than I thought; I hope that it somehow makes sense, otherwise I'm very happy about suggestions!

**New features:**
- Added the ability to use integers of any size as bit vectors (such as `u128`). I'd like to default to `u64`, but haven't yet worked out how to do it correctly (see open questions).
- Added two new constructors which allow dealing with more complex ambiuities ('wildcards') (`with_ambig`, `with_ambig_results`).
- Added a traceback algorithm comparable to the one of Edlib (https://github.com/Martinsos/edlib), allowing to obtain the starting positions of a match. There is a new `find_all_pos` method, which takes a mutable reference to self and returns an Iterator over tuples of start, end and distance. The traceback is faster than the one of Edlib. The latter is probably slowed down by the handling longer patterns. I also replaced a loop with `count_ones()`, which speeds things up, at least on x86 (not sure about ARM).

  Another method (`find_all_pos_remember`) allows searching the text and obtaining a traceback *later* at any position in the range which has already been searched. This uses more memory since the states are stored for the whole search text, not only the length of the pattern. However, the feature can be useful e.g. if only the traceback of the best hit is needed. The type returned by `find_all_pos_remember` is `FullMatches<..., Remember>`. `Remember` is a 'marker' struct which ensures that methods for obtaining the traceback at any position (`hit_at`, `path_at`, `alignment_at`) are only available if all states are remembered. I'm not sure whether this works with any Rust version and makes sense. I could also use a macro for implementing two different types of `FullMatches`.

  The kind of information returned from the traceback can be configured. Since there are performance differences, I provided different methods for all options.
  - Just the starting position of the hit (fastest; using the `FullMatches` iterator or `FullMatches::start()`, or `hit_at()`).
  - Alignment path (`FullMatches::path()` / `next_path()` / `path_at()`); slightly slower.
  - Update an `alignment::Alignment` instance with the path and all positional information (`FullMatches::alignment()` / `next_alignment()` / `alignment_at()`). This is again slighly slower than just obtaining the aligment path

  A method *returning* an `Alignment` instance is not provided on purpose since creating a new instance each time has a major performance impact, and therefore I'm trying to enforce reusing the instances if possible. To make life easier, a constructor for 'empty' alignments has been added to `Alignment`. The `AlignmentMode` struct has `Default` implemented, (defaults to `AlignmentMode::Custom`).

- I did some benchmarks living in `benches/approximate_matching.rs`. Using LTO can have unpredictable effects and lead to slowdowns. However, adding `codegen-units = 1` helps, although the speedup over not using LTO is mostly marginal.

**Open questions:**
- Does the API of `with_ambig` and `with_ambig_results` make sense? How about the function names?
- The current traceback algorithm follows Edlib, where gaps in the search text are preferred over gaps in the search pattern. A semiglobal alignment using `alignment::pairwise::Aligner` seems to do it the other way round. I personally prefer the current implementation, but maybe there are good reasons against it.
- Should I differentiate `FullMatches` into two separate types with partially identical methods instead of using generics and the `(No)Remember` types? (could be done with a macro)
- Should the `Iterator` implementation of `FullMatches` return the starting positions as well (as currently done), or should it behave exactly as `Matches`, returning only end position and distance? I think the first option is what I whould expect from a type with the name `FullMatches`. The second option has the advantage that the expensive traceback is avoided by default and only done if really needed using the hit_at / alignment_at functions. I chose the first option since it seems more intuitive
- Is it a good idea to implement `Default` for `AlignmentMode`? This is not strictly necessary, there could also be an `Alignment` constructor which chooses a default, while `AlignmentMode` is not touched.
- A general Rust question: is it possible to have `Myers::new()` default to `Myers<u64>`, thus be a synonym of `Myers64::new()`? Currently, the compiler wants type annotations despite of the `T = u64` in the struct definition.
- In the `Traceback::add_state()` method, unsafe is used due to performance reasons. I hope this is ok.

**Other things not yet done:**
- Write even more tests covering every use case
- Update module-level docs with all relevant information

markschl commented:
I'm not sure why 'Fix for #117 ' appears in the list, it's from a different pull request and nothing was changed in ukkonen.rs

markschl commented:
Results of a benchmark (see benches/approximate_matching.rs), run on a Mid 2010 Mac Pro:

```
test myers_alignment_64                ... bench:     224,836 ns/iter (+/- 418)
test myers_alignment_alloc_64          ... bench:     219,756 ns/iter (+/- 262)
test myers_best_alignment_64           ... bench:     155,616 ns/iter (+/- 279)
test myers_best_alignment_itertools_64 ... bench:     241,253 ns/iter (+/- 300)
test myers_end_128                     ... bench:     196,538 ns/iter (+/- 113)
test myers_end_32                      ... bench:     100,521 ns/iter (+/- 110)
test myers_end_64                      ... bench:      99,272 ns/iter (+/- 57)
test myers_ends_noremember_64          ... bench:     143,169 ns/iter (+/- 156)
test myers_ends_remember_64            ... bench:     144,198 ns/iter (+/- 230)
test myers_path_64                     ... bench:     173,488 ns/iter (+/- 682)
test myers_pos_128                     ... bench:     315,693 ns/iter (+/- 1,560)
test myers_pos_64                      ... bench:     169,899 ns/iter (+/- 923)
test ukkonen                           ... bench:   2,234,366 ns/iter (+/- 10,946)
```

The benchmark `myers_end_64` corresponds to the current implementation. The Ukkonen algorithm was added for comparison.

markschl commented:
Another comment: I forgot to highlight that this pull request actually changes the existing API: the `max_dist` parameter and the distance returned after searching were switched from `u8` to `usize`. This is not actually required with 128-bit integers. But since the `Num` trait could also be implemented for any big-integer type, I changed it. Alternatively, it could be a good idea to have an associated type of the `Num` trait for specifying the size of the distance integer.

johanneskoester commented on src/pattern_matching/myers.rs:
```diff
@@ -13,75 +13,246 @@
+    /// }));
+    /// assert_eq!(myers_ambig_symm.distance(text), 0);
+    /// # }
+    pub fn with_ambig<P, I, C>(pattern: P) -> Self
```
I think it could be easier to handle this with a builder pattern (best using the [builder crate](https://docs.rs/derive_builder/0.5.1/derive_builder/)). Something like

```
Myers::new(pattern).ambig(b'N', b"ACGT").build();
```

markschl replied:
Ah yes, that's a good idea, seems much simpler. However, I'm not sure how to implement this. The pattern is not stored directly, but 'encoded' in the `peq` array. I guess it would need to be stored in a separate `Vec`. Still, after each call to `.ambig()`, we would have to search the pattern for occurrences of this character. 

I guess in most cases the negative performance impact is neglectable since the `Myers` instance may be constructed once and used to search many sequences. Still, I don't know.


One possibility for simplification of my current implementation came into my mind: How about a trait similar to [std::str::pattern](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html), which would allow to specify either an `IntoIterator<Item=u8>`, but also nested iterators. I'm not sure whether that could easily be implemented, I would have to try first. Of course, it's still not as ergonomic as the builder pattern.

johanneskoester replied:
I think I still prefer the builder. The builder struct itself can store each ambig declaration in a Vec. Then, upon build you can parse the pattern and perform the replacement. This way, multiple calls to ambig do not create the kind of overhead you describe, because the pattern only has to be converted once.

markschl replied:
Now I get it, yes that sounds very good. Maybe, this `MyersBuilder` (or how it would be called) could still be optional (just for the case of adding ambiguities) and the `Myers::new` method could be preserved.

johanneskoester replied:
Yes, I agree.

johanneskoester commented on src/pattern_matching/myers.rs:
```diff
@@ -13,75 +13,246 @@
 
     /// Create a new instance of Myers algorithm for a given pattern and a wildcard character
     /// that shall match any character.
     pub fn with_wildcard(pattern: TextSlice, wildcard: u8) -> Self {
```
Can also go into the builder pattern.

johanneskoester commented on src/pattern_matching/myers.rs:
```diff
@@ -105,47 +282,160 @@ impl Myers {
+    }
+
+    /// Like `find_all_pos`, but additionally allows for obtaining the starting positions and/or
+    /// the alignment at *any* position that was already searched.
```
I don't really get this.

johanneskoester commented on src/pattern_matching/myers.rs:
```diff
@@ -156,29 +446,374 @@ impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
 
+// 'Marker' traits
+pub struct Remember;
+pub struct NoRemember;
```
Can you explain the purpose in the comment above? Would an enum make sense for the two values?

markschl replied:
'Traits' is completely wrong, I should have written 'structs'. The term 'remember' may not be the best word, in general the idea is that with `NoRemember`, only states needed for obtaining the alignment path at the *currently searched* position are stored. With `Remember`, states at each position in the text are stored. Consequently, the path can later be obtained for *any* position that was already searched (using `alignment_at` and related methods)

johanneskoester commented on src/pattern_matching/myers.rs:
```diff
@@ -156,29 +446,374 @@ impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+}
+
+
+struct Traceback<T: Num> {
```
The traceback for Myers is very interesting. Have you invented that yourself or is it from a textbook? Could you add details about the complexity implications? How is it configured?

markschl replied:
I experimented with different implementations, but ended up doing it [similarly as Edlib does](https://github.com/Martinsos/edlib/blob/6967b806275cd4766ed6c58342a0dff8509e70ad/edlib/src/edlib.cpp#L924). Of course, the implementation was much simpler because I didn't have to care about blocks, and only semiglobal alignment possible, while Edlib also does global and local alignments. The author states that his implementation is inspired by [SSWL](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0082138), but I haven't looked into this. I was searching for other implementations, also in SeqAn, but didn't find any, though it's possible that there is one.

There is some description in [this master thesis; chapter 4.2.2 (pg. 32)](https://bib.irb.hr/datoteka/758607.diplomski_Martin_Sosic.pdf). There is also a [paper about Edlib](https://academic.oup.com/bioinformatics/article/33/9/1394/2964763), but there I couldn't find any details about the implementation.

Finding the score in the i-th cell always requires checking all bits starting from the last cell down to the i-th cell. The simples approach is to use a loop, but I found that using popcnt (`count_ones()`) was faster, so that's my only contribution :).

`_traceback_at()` starts at the given end position (which should have been searched, otherwise the result will be nonsense) and iterates backwards, finding the scores of individual cells and the best alignment path.

Due to a lack of theoretical knowledge, I'm not completely sure about the complexity implications. There is one loop going from zero to `m` (pattern length), so I guess time complexity could be O(m). Space complexity depends on whether all states are stored ('remembered') or not. In the first case, I guess it's O(n + k), in the second case O(m + k).

I hope this answers your questions, otherwise I'm happy to give more details.

johanneskoester replied:
Yes, thanks! Please add the used literature to the doc comment.

markschl commented on src/pattern_matching/myers.rs:
```diff
@@ -13,75 +13,246 @@
+    fn wrapping_add(&self, other: &Self) -> Self;
+}
+
+// Implementation using num_traits types
```
The whole idea of the `Num` trait (or however it may be called) was to enable custom implementations that don't depend on the num_traits crate, with big integers in mind. The problem is that only a single method is needed from `num_traits::int::PrimInt`, but if depending on PrimInt directly, a whole bunch of unneeded methods would need to be implemented.
However, there are other hurdles for the use of big-integer types (wrapping_add seems missing in all crates), and I have no idea about how performance would be. Therefore I'm totally ok with removing this trait and directly depending on the num_traits types.

johanneskoester replied:
Yes, lets do that for now. I think it would be preferable to fix that in the num_traits crate sooner or later.

markschl commented on src/pattern_matching/myers.rs:
```diff
@@ -156,29 +446,374 @@ impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+            o.clear();
+        }
+
+        // Mask with one bit switched on at the current position
```
This should actually be like "Bit mask covering all positions of the pattern"

markschl commented:
Regarding the searching API, I assembled a few examples that illustrate how it currently works and what are my ideas for further modification.

```rust

let text = b"ACCGTGGATGAGCGCCATAG";
let pattern =      b"TGAGCGT";

let mut myers = Myers64::new(pattern);

// find all end positions and distances (`Matches` type)
// This is already possible in rust-bio
for (end, dist) in myers.find_all_end(text, 1) {
    println!("end: {}, dist: {}", end, dist);
}
// end: 13, dist: 1
// end: 14, dist: 1


// Determine starting positions as well
// (the type is currently: `FullMatches<..., NoRemember>`)
for (start, end, dist) in myers.find_all_pos(text, 1) {
    println!("{} .. {}, dist: {}", start, end, dist);
}
// note that 'end' is now a range end coordinate and thus
// not the same as the 0-based index:
// 8 - 14, dist: 1
// 8 - 15, dist: 1


// In order to get an alignment:

let mut aln = Alignment::new(); // or Alignment::default()?

{ // separate scope needed to make the borrow checker happy (examples below would not work)
    let mut hits = myers.find_all_pos(text, 1);
    while hits.next_alignment(&mut aln) {
        println!(
            "{} .. {}, dist: {}\n{}", 
            aln.ystart, aln.yend, aln.score, aln.pretty(pattern, text)
        );
    }
    // 8 .. 14, dist: 1
    //        TGAGCGT
    //        ||||||+
    // TGAGCGTTGAGCG-
    //
    // 8 .. 15, dist: 1
    //        TGAGCGT
    //        ||||||\
    // TGAGCGTTGAGCGC
}


// Now, where it becomes a bit more complicated is
// if the alignment path is determined *later*
{
    // The type is currently: `FullMatches<..., Remember>`
    let mut hits = myers.find_all_pos_remember(text, 1);
    // the following code could be simplified using itertools,
    // see
    let mut best_dist = ::std::usize::MAX;
    let mut best_end = 0;
    while let Some((end, dist)) = hits.next_end() {
        if dist < best_dist {
            best_dist = dist;
            best_end = end;
        }
    }

    hits.alignment_at(best_end, &mut aln);
    println!(
        "The best hit is: {} .. {}, dist: {}\n{}",
        aln.ystart, aln.yend, aln.score, aln.pretty(pattern, text)
    );
    // // Well, there is no single 'best hit' here, it's just the first 
    // // of two hits with the same distance
    // The best hit is: 8 .. 14, dist: 1
    //        TGAGCGT
    //        ||||||+
    // TGAGCGTTGAGCG-
}


// The above example is very complicated. How about using a kind of builder pattern
// here, too: (not currently working)
{
    let mut hits = myers
        .find_all_end(text, 1)
        // This is new. The name of the function is probably weird, but
        // I can't think of a better one right now. The type
        // could be `MatchesRemember` following the above convention
        .remember_states();

    let (best_end, best_dist) = (&mut hits)
        .min_by_key(|&(_, dist)| dist)
        .unwrap();

    hits.alignment_at(best_end, &mut aln);
    // as above...
}
```

Following the last example, there could be 4 different versions of `Matches`:
- `find_all_end()` -> `Matches`
- `find_all_pos()` -> `FullMatches<..., NoRemember>` or just `FullMatches`
- `find_all_end().remember_states()` -> `Matches<..., Remember>` or `MatchesRemember`
- `find_all_pos().remember_states()` -> `FullMatches<..., Remember>` or `FullMatchesRemember`
As stated above, 'remember' is probably not the best word...

There's just one thing: `find_all_end()` takes `self` by immutable reference, but storing the states in `Traceback` is only possible with a mutable reference. Therefore, `find_all_end().remember_states()` does not work unless `find_all_end()` takes self by mutable reference as well.

I hope that this API would not be too complicated. I personally like the idea of being able to do a traceback *later* at *any* position.

markschl commented on src/pattern_matching/myers.rs:
```diff
@@ -156,29 +446,374 @@ impl<'a, I: Iterator<Item = &'a u8>> Iterator for Matches<'a, I> {
+        }
+
+        // Mask with one bit switched on at the current position
+        let max_mask = T::one() << (self.m - 1);
```
This should actually be like "Bit mask covering all positions of the pattern"

johanneskoester replied:
Ok, please add that to the comment.

johanneskoester commented:
Thanks for the explanation of the API. I think the builder pattern is not needed. However, what about renaming this remember thing to `lazy`. To me this seems to be more intuitive because it already explains the behavior.

johanneskoester commented:
If not done yet, please add all the nice example above to the doc comment of the module, but separated into sections with Markdown.

johanneskoester commented on src/pattern_matching/myers/builder.rs:
```diff
@@ -0,0 +1,179 @@
+/// let mut builder = MyersBuilder::new();
+///
+/// for &(base, equivalents) in &ambigs {
+///     builder.ambig(base, equivalents);
```
I would probably write this directly in this form:
```
builder.ambig(b'M', b"ACM");
builder.ambig(b'R', b"AGR");
```

Also, it feels unintuitive that the letter itself is repeated in the pattern.

markschl replied:
I don't entirely get the above point. Do you mean the usage of `&b"ACM"[..]`? It's not possible to do it using simply `b"ACM"` since these byte arrays are of different length.
About repeating the letter itself in the pattern: agreed! I changed this behavior.

johanneskoester commented on fuzz/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
```
Fuzzing, great idea! I wasn't aware of cargo-fuzz before.

markschl commented:
Thanks for the review! After few additional commits I consider the code now ready for final review. 

Here are some random points and questions:

* Fixed two bugs in the existing algorithm (commits d442aece6302b397cd917b9bf4e2024713c5da84 and a94b137cc50ad295f214c5e6999f704593abd080)
* I haven't figured out whether there is a way to call Myers::new() without explicit type annotation and have it default to u64. Therefore, Myers64::new() is always used to construct Myers<u64>. Still, I think if referencing &Myers without type annotation, Myers<u64> will be assumed. There is a bit of an inconsistency in the API, as MyersBuilder::build() returns Myers<u64>, and the generic method is called build_other(). Maybe, build() should be renamed to build64().
* I hope that the terminology (find_all vs. find_all_end and find_all_lazy makes sense).
* Should Alignment::score be negative as in pairwise alignment? Up to now, it's positive, just representing the edit distance
* I added a new_alignment() method, which creates an 'empty' `Alignment` that may be reused. However, I'd be happy about a constructor in bio_types, which can be used here. The only problem I see there: which value should be chosen for `AlignmentMode`? Do you have any thoughts on that? Regarding the benefits of reusing Alignment you may compare the benchmarks `myers_alignment_64` and `myers_alignment_alloc_64`. On the systems I tested, there was definitely a difference.

johanneskoester commented on src/pattern_matching/myers/mod.rs:
```diff
@@ -0,0 +1,980 @@
+pub type Myers128 = Myers<u128>;
+
+/// Myers algorithm.
+pub struct Myers<T>
```
I think you can solve your generics issue by defining
```
pub struct Myers<T=u64>
```

johanneskoester commented on src/pattern_matching/myers/builder.rs:
```diff
@@ -0,0 +1,187 @@
+    }
+
+    /// Creates a Myers instance given a pattern, using `u64` as bit vector type
+    pub fn build<'a, P>(&self, pattern: P) -> Myers<u64>
```
I would rename build_other into this, and have two shorthands `build_64` and `build_128`.

markschl commented:
Thanks for adding `Alignment::default`. I've adjusted the code to use bio_types >= 0.4. I've also changed the method names of `MyersBuilder`. Actually, it came to my mind that it might also just be called `AmbigBuilder`, since configuring ambiguities is the only task it does. On the other hand there may be some future use that I'm not foreseeing...

`u64` is now the default bitvector (`struct Myers<T = u64>`). In struct and function definitions, not writing the type argument will thus work as before. But `Myers::new` still requires type annotations and will thus break existing code. I assume that it is not possible for `Myers::new` to serve both as generic constructor and as constructor of `Myers<u64>`. `Myers64::new()` should be used instead. 
- Alternatively the new conventions of `MyersBuilder` could be followed, using `Myers::new_64()` instead. 
- Or, we could default to `u64` for `Myers::new()`. For consistency, this would involve reverting the last API changes to MyersBuilder.

johanneskoester commented:
I think the breaking change is ok. I would also say that we don't need Myers64 at all, because it is no win over `Myers<u64>`.

markschl commented:
Removing this type requires to be much more verbose in many of the documentation examples:
```
let myers: Myers<u64> = Myers::new();
```
instead of 
```
let myers = Myers64::new();
```
That's the reason I added it.

johanneskoester commented:
You can write `Myers::<u64>::new()`.

markschl commented:
However, regarding build_64 and build_128, they require two type parameters (P for pattern as well), thus we have to write `MyersBuilder::new()....build::<u64, _>(&pattern)`. Maybe, preserving those is worth thinking about. Of course, there's also `let myers: Myers<u64> = MyersBuilder::new()....build(&pattern)`

markschl commented:
Done...
